src/core/border_router/nat64_translator.cpp

```diff
@@ -0,0 +1,454 @@
+    }
+
+    if (!mNat64Prefix.IsValidNat64())
+    {
+        LogWarn("incoming message is an IPv4 packet but no NAT64 prefix configured, drop");
+        ExitNow(res = Result::kDrop);
+    }
+
+    if (ip4Header.GetTtl() <= 1)
+    {
+        LogDebg("incoming packet TTL reached");
+        ExitNow(res = Result::kDrop);
+    }
+    ip4Header.SetTtl(ip4Header.GetTtl() - 1);
+
+    mapping = GetMapping(ip4Header.GetDestination());
+    if (mapping == nullptr)
+    {
+        LogWarn("no mapping found for the IPv4 address");
+        ExitNow(res = Result::kDrop);
+    }
+
+    aMessage.RemoveHeader(sizeof(Ip4::Header));
+
+    ip6Header.Clear();
+    ip6Header.InitVersionTrafficClassFlow();
+    ip6Header.GetSource().SynthesizeFromIp4Address(mNat64Prefix, ip4Header.GetSource());
+    ip6Header.SetDestination(mapping->mIp6);
+    ip6Header.SetFlow(0);
+    ip6Header.SetHopLimit(ip4Header.GetTtl());
+
+    // Note: TCP and UDP are the same for both IPv4 and IPv6 except for the checksum calculation, we will update the
+    // checksum in the payload later. However, we need to translate ICMPv6 packets to ICMP in IPv4 packets.
+    switch (ip4Header.GetProtocol())
+    {
+    case Ip4::kProtoUdp:
+        ip6Header.SetNextHeader(Ip6::kProtoUdp);
+        res = Result::kForward;
+        break;
+    case Ip4::kProtoTcp:
+        ip6Header.SetNextHeader(Ip6::kProtoTcp);
+        res = Result::kForward;
+        break;
+    case Ip4::kProtoIcmp:
+        ip6Header.SetNextHeader(Ip6::kProtoIcmp6);
+        SuccessOrExit(TranslateIcmp4(*mapping, aMessage));
+        res = Result::kForward;
+        break;
+    default:
+        ExitNow(res = Result::kDrop);
+    }
+
+    switch (res)
+    {
+    case Result::kDrop:
+        break;
+    case Result::kReplyIcmp:
+        // TODO: Implement the logic for replying ICMP packets.
+        break;
+    case Result::kForward:
+        ip6Header.SetPayloadLength(aMessage.GetLength() - aMessage.GetOffset());
+        Checksum::UpdateMessageChecksum(aMessage, ip6Header.GetSource(), ip6Header.GetDestination(),
+                                        ip6Header.GetNextHeader());
+        err = aMessage.PrependBytes(&ip6Header, sizeof(ip6Header));
+        if (err != kErrorNone)
+        {
+            // This might happen when the platform failed to reserve enough space before the original IPv4 packet.
+            LogWarn("failed to prepend IPv6 head to translated message");
+            res = Result::kDrop;
+        }
+        break;
+    }
+
+exit:
+    return res;
+}
+
+void Nat64Translator::ReleaseMapping(AddressMapping &aMapping)
+{
+    LogInfo("mapping removed: %s -> %s", aMapping.mIp6.ToString().AsCString(), aMapping.mIp4.ToString().AsCString());
+    mIp4AddressPool[mAvailableAddressCount] = aMapping.mIp4;
+    mAvailableAddressCount++;
+    mAddressMappingPool.Free(aMapping);
+}
+
+Nat64Translator::AddressMapping *Nat64Translator::CreateMapping(const Ip6::Address &aAddr)
```

Comment:
Use `aIp6Addr` instead of `aAddr` for param name (helps make it clear it is IPv6 address)?